funsor/distributions.py

```diff
@@ -88,69 +91,111 @@ def eager_reduce(self, op, reduced_vars):
-        value = to_funsor(value, reals())
-        return probs, value
+    @makefun.with_signature(f"__init__(self, {', '.join(param_names)}, value='value')")
+    def dist_init(self, **kwargs):
+        return Distribution.__init__(self, *tuple(kwargs[k] for k in self._ast_fields))
 
-    def __init__(self, probs, value=None):
-        super(BernoulliProbs, self).__init__(probs, value)
+    dist_class = DistributionMeta(pyro_dist_class.__name__.split("__")[-1], (Distribution,), {
+        'dist_class': pyro_dist_class,
+        '__init__': dist_init,
+    })
 
+    eager.register(dist_class, *((Tensor,) * (len(param_names) + 1)))(dist_class.eager_log_prob)
 
-@eager.register(BernoulliProbs, Tensor, Tensor)
-def eager_bernoulli(probs, value):
-    return BernoulliProbs.eager_log_prob(probs=probs, value=value)
+    return dist_class
 
 
-class BernoulliLogits(Distribution):
-    """
-    Wraps :class:`pyro.distributions.Bernoulli` .
+class __BernoulliProbs(dist.Bernoulli):
+    def __init__(self, probs, validate_args=None):
+        return super().__init__(probs=probs, validate_args=validate_args)
 
-    :param Funsor logits: Log likelihood ratio of 1.
-        This should equal ``log(p1 / p0)``.
-    :param Funsor value: Optional observation in ``{0,1}``.
-    """
-    dist_class = dist.Bernoulli
 
-    @staticmethod
-    def _fill_defaults(logits, value='value'):
-        logits = to_funsor(logits)
-        assert logits.dtype == "real"
-        value = to_funsor(value, reals())
-        return logits, value
+class __BernoulliLogits(dist.Bernoulli):
+    def __init__(self, logits, validate_args=None):
+        return super().__init__(logits=logits, validate_args=validate_args)
+
+
+class __CategoricalLogits(dist.Categorical):
+    def __init__(self, logits, validate_args=None):
+        return super().__init__(logits=logits, validate_args=validate_args)
+
 
-    def __init__(self, logits, value=None):
-        super(BernoulliLogits, self).__init__(logits, value)
+_wrapped_pyro_dists = [
+    (dist.Beta, ()),
+    (__BernoulliProbs, ('probs',)),
+    (__BernoulliLogits, ('logits',)),
+    (dist.Binomial, ('total_count', 'probs')),
+    (dist.Multinomial, ('total_count', 'probs')),
+    (dist.Categorical, ('probs',)),
+    (__CategoricalLogits, ('logits',)),
+    (dist.Poisson, ()),
+    (dist.Gamma, ()),
+    (dist.VonMises, ()),
+    (dist.Dirichlet, ()),
+    (dist.DirichletMultinomial, ()),
+    (dist.Normal, ()),
+    (dist.MultivariateNormal, ('loc', 'scale_tril')),
+    (dist.Delta, ()),
+]
+
+for pyro_dist_class, param_names in _wrapped_pyro_dists:
+    locals()[pyro_dist_class.__name__.split("__")[-1].split(".")[-1]] = make_dist(pyro_dist_class, param_names)
+
+# Delta has to be treated specially because of its weird shape inference semantics
+Delta._infer_value_domain = classmethod(lambda cls, **kwargs: kwargs['v'])
+
+
+# Multinomial and related dists have dependent bint dtypes, so we just make them 'real'
```

Comment:
Good point, would you mind pointing to this new issue so we can start planning refactoring?
https://github.com/pyro-ppl/funsor/issues/322